ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp

```diff
@@ -351,4 +372,104 @@ void screen_redraw()
+  draw_batch_flush(batch_flush_deferred);
+
+  bool flipped = false;
+  unsigned char* rgba = enigma::graphics_copy_screen_pixels(x,y,w,h,&flipped);
+  int ret = image_save(filename, rgba, w, h, w, h, flipped);
+
+  delete[] rgba;
+  return ret;
+}
+
+int background_create_from_screen(int x, int y, int w, int h, bool removeback, bool smooth, bool preload)
+{
+  draw_batch_flush(batch_flush_deferred);
+
+  bool flipped = false;
+  unsigned char* rgba = enigma::graphics_copy_screen_pixels(x,y,w,h,&flipped);
+
+  if (flipped)
+    rgba = enigma::image_flip(rgba, w, h, 4);
+
+  enigma::backgroundstructarray_reallocate();
+  int bckid=enigma::background_idmax;
+  enigma::background_new(bckid, w, h, &rgba[0], removeback, smooth, preload, false, 0, 0, 0, 0, 0, 0);
+  delete[] rgba;
+  enigma::background_idmax++;
+  return bckid;
+}
+
+int sprite_create_from_screen(int x, int y, int w, int h, bool removeback, bool smooth, bool preload, int xorig, int yorig) {
+  draw_batch_flush(batch_flush_deferred);
+
+  bool flipped = false;
+  unsigned char* rgba = enigma::graphics_copy_screen_pixels(x,y,w,h,&flipped);
+
+  if (flipped)
+    rgba = enigma::image_flip(rgba, w, h, 4);
+
+  enigma::spritestructarray_reallocate();
+  int sprid=enigma::sprite_idmax;
+  enigma::sprite_new_empty(sprid, 1, w, h, xorig, yorig, 0, h, 0, w, preload, smooth);
+  enigma::sprite_set_subimage(sprid, 0, w, h, rgba, rgba, enigma::ct_precise); //TODO: Support toggling of precise.
+  delete[] rgba;
+  return sprid;
+}
+
+int sprite_create_from_screen(int x, int y, int w, int h, bool removeback, bool smooth, int xorig, int yorig) {
+  return sprite_create_from_screen(x, y, w, h, removeback, smooth, true, xorig, yorig);
+}
+
+void sprite_add_from_screen(int id, int x, int y, int w, int h, bool removeback, bool smooth) {
+  draw_batch_flush(batch_flush_deferred);
+
+  bool flipped = false;
+  unsigned char* rgba = enigma::graphics_copy_screen_pixels(x,y,w,h,&flipped);
+
+  if (flipped)
+    rgba = enigma::image_flip(rgba, w, h, 4);
+
+  enigma::sprite_add_subimage(id, w, h, rgba, rgba, enigma::ct_precise); //TODO: Support toggling of precise.
+  delete[] rgba;
+}
+
+// The following two leave a bad taste in my mouth because they depend on views, which should be removable.
+// However, for now, they stay.
+
+int draw_getpixel(int x,int y)
+{
+  if (clamp_view(x,y)) return 0;
+  draw_batch_flush(batch_flush_deferred);
+
+  unsigned char* rgba = enigma::graphics_copy_screen_pixels(x,y,1,1);
+  int ret = rgba[2] | rgba[1] << 8 | rgba[0] << 16;
+  delete[] rgba;
+  return ret;
+}
+
+int draw_getpixel_ext(int x,int y)
+{
+  if (clamp_view(x,y)) return 0;
+  draw_batch_flush(batch_flush_deferred);
+
+  unsigned char* rgba = enigma::graphics_copy_screen_pixels(x,y,1,1);
+  int ret = rgba[2] | rgba[1] << 8 | rgba[0] << 16 | rgba[3] << 24;
+  delete[] rgba;
+  return ret;
+}
+
```

Comment:
Here's another problem you can solve (feel free to do so later or not at all, but if the latter, please file a bug):

`draw_getpixel`: legacy function; calls `draw_getpixel_fresh()`.
`draw_getpixels`: populates a buffer with pixel data from the screen and retrieves the pixel at the specified coordinates. FUTURE: returns the fetched pixel data as a flat_matrix.
`draw_getpixel_cached`: returns the most-recently-read pixel cache at the given coordinate.

Reply:
That's a good idea, I actually like that. Originally I had thought of `draw_getpixels` just returning a var array. Here's the real problem though, whacky people used this for heightmap reading, which including drawing the texture to the screen first. I dislike all of this, because of what I did now, we could just add texture_get_pixels which would cover sprites and backgrounds and surfaces (just not the screen). Though, I suppose there's a few useful cases yet for draw_getpixel (hit detection).